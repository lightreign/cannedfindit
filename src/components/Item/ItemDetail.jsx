import React from "react";
import { connect } from 'react-redux';
import { Alert } from "react-bootstrap-v5";
import {Link} from "react-router-dom";

const ItemDetail = ({item}) => {
    return (
        <div>
            {item ?
            <div>
                <h4>Item: {item.product.brand.name} {item.product.type.name}</h4>
                <p>Location: {item.location.name}</p>
                <p>Expiry: {item.expiry.toDateString()}</p>
            </div>
                : <Alert variant="warning">
                    Item could not be found, please go to <Link to="/">Dashboard</Link>
                </Alert> }
        </div>
    );
};

const mapStateToProps = (state, ownProps) => {
   let id = ownProps.match.params.id;
   let item = state.items.find(item => item._id === id);

   return {
       item: item
   };
}

export const ConnectedItemDetail = connect(mapStateToProps)(ItemDetail);
