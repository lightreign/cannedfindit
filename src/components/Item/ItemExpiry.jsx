import React from "react";

export const ItemExpiry = ({item}) => {
    let expiryClass = '';

    if (item.isExpired()) {
        expiryClass = 'text-danger';
    } else if (item.expiringVerySoon()) {
        expiryClass = 'text-warning';
    } else if (item.expiringSoon()) {
        expiryClass = 'text-info';
    }

    return (
        <div className={expiryClass}>{item.expiryAsString()}</div>
    );
}

