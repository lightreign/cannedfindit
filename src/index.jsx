import "core-js/stable";
import "regenerator-runtime/runtime";
import React from 'react';
import ReactDOM from 'react-dom';
import 'react-datetime/css/react-datetime.css';
import './resources/sass/app.scss';
import { App } from "./components/App";

ReactDOM.render(
    <App/>,
    document.getElementById('mountNode')
);
