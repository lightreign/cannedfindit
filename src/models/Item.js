const moment = require('moment');

/**
 * Item model
 */
class Item {
    constructor(item) {
        for (let key in item) {
            // eslint-disable-next-line no-prototype-builtins
            if (item.hasOwnProperty(key)) {
                this[key] = item[key];
            }
        }

        this.expiry = new Date(this.expiry);

        if (this.consumed) {
            this.consumed = new Date(this.consumed);
        }
    }

    expiryDateString() {
        return this.expiry.toLocaleDateString();
    }

    consumedDateString() {
        return this.consumed ? this.consumed.toLocaleDateString() : null;
    }

    isExpired() {
        return this.expiry <= moment();
    }

    expiringVerySoon() {
        return this.expiry <= moment().add(1, 'month');
    }

    expiringSoon() {
        return this.expiry <= moment().add(3, 'month');
    }
}

export default Item;