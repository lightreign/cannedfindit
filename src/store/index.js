import { createStore, combineReducers, applyMiddleware } from "redux";
import { composeWithDevTools } from "redux-devtools-extension";
import thunk from 'redux-thunk';
import { defaultState } from "../server/defaultState";
import { itemReducer, typeReducer } from "./reducers";

export const store = createStore(
    combineReducers({
        user(user = defaultState.user) {
            return user;
        },
        items: itemReducer,
        types: typeReducer
    }),
    composeWithDevTools(applyMiddleware(thunk))
);
