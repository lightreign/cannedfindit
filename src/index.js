import "core-js/stable";
import "regenerator-runtime/runtime";
import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from "react-redux";
import { BrowserRouter as Router, Route } from "react-router-dom";
import bootstrap from 'bootstrap';
import 'bootstrap/dist/css/bootstrap.min.css';

import App from './components/App';
import { ConnectedDashboard } from "./components/Dashboard";
import { ConnectedNavigation } from "./components/Navigation";
import { ConnectedItemDetail } from "./components/ItemDetail";
import { inventory } from "./inventory";

ReactDOM.hydrate(
    <Router>
        <Provider store={inventory}>
            <ConnectedNavigation/>
            <Route exact path="/btngame" component={App}/>
            <Route exact path="/" component={ConnectedDashboard}/>
            <Route exact path="/item/:id" render={({match}) => (<ConnectedItemDetail match={match}/>)}/>
        </Provider>
    </Router>,
    document.getElementById('mountNode')
);
