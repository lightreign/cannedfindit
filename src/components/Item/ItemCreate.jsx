import React from "react";
import Datetime from 'react-datetime';
import { addItem } from "../../store/actions";
import { connect } from "react-redux";
import { ConnectedLocationSelect } from "../Location/LocationSelect";
import { ConnectedProductSelect } from "../Product/ProductSelect";
import { Button, Form } from "react-bootstrap-v5";

export const ItemCreate = ({createItem, setProduct, setExpiry, setLocation}) => (
    <Form id="itemCreateForm" onSubmit={createItem}>
        <legend>Create an Item</legend>

        <ConnectedProductSelect setProduct={setProduct}/>

        <Form.Group controlId="expiry">
            <Form.Label>Expiry:</Form.Label>
            <Datetime dateFormat="DD/MM/YYYY" onChange={setExpiry} timeFormat={false} closeOnSelect={true}/>
        </Form.Group>

        <ConnectedLocationSelect setLocation={setLocation}/>

        <Button variant="primary" type="submit">Add Item</Button>
    </Form>
);

const mapStateToProps = (state, ownProps) => {
    return {
        user: state.user
    };
}

const mapDispatchToProps = (dispatch, ownProps) => {
    return {
        createItem(e) {
            e.preventDefault();

            if (ownProps.product && ownProps.location && ownProps.expiry) {
                const item = {
                    product: ownProps.product,
                    location: {
                        name: ownProps.location
                    },
                    expiry: ownProps.expiry
                };

                dispatch(addItem(item));
                e.target.reset();
            } else {
                // TODO: error message
            }
        },
        setProduct(e) {
            ownProps.product = JSON.parse(e.target.value);
        },
        setExpiry(date) {
            ownProps.expiry = date.toDate();
        },
        setLocation(e) {
            ownProps.location = e.target.value;
        }
    };
};

export const ConnectedItemCreate = connect(mapStateToProps, mapDispatchToProps)(ItemCreate);