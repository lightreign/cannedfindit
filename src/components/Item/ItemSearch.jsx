import React, { useState } from "react";
import PropTypes from "prop-types";
import { Button, Form } from "react-bootstrap-v5";
import { connect } from "react-redux";
import { listItems } from "../../store/actions";

export const ItemSearch = ({dispatch, changeMode}) => {
    const [productType, setProductType] = useState('');
    const [searched, setSearched] = useState(true);

    const searchItems = (e) => {
        e.preventDefault();

        // A little hacky, if no product type reset form
        // TODO: This will get addressed in controlled input refactor
        if (!productType.length) {
            e.target.reset();
        }

        setSearched(true);

        changeMode('list');

        dispatch(listItems({'product.type.name': productType}, 1));
    }

    const searchChange = e => {
        setSearched(false);
        setProductType(e.target.value);
    };

    const onClear = () => {
        setSearched(false);
        setProductType('');
    };

    return (
        <Form id="searchForm" onSubmit={searchItems} role="SearchForm">
            <Form.Group controlId="searchItemProductType">
                <Form.Label>Search</Form.Label>
                <Form.Control name="search" onChange={searchChange} data-testid="search"/>
            </Form.Group>

            <Button variant="primary" type="submit" disabled={searched}>Search</Button>
            <Button variant="secondary" type="submit" onClick={onClear}>Clear</Button>
        </Form>
    );
};

ItemSearch.propTypes = {
    dispatch: PropTypes.func.isRequired,
    changeMode: PropTypes.func.isRequired
};

const mapStateToProps = (state) => {
    return {
        user: state.user,
        items: state.items,
        pager: state.pager,
    };
}

export const ConnectedItemSearch = connect(mapStateToProps)(ItemSearch);

