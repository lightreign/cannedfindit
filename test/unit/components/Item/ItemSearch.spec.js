import React from "react";
import renderer from 'react-test-renderer';
import { Provider } from "react-redux";
import { ItemSearch } from "../../../../src/components/Item/ItemSearch";
import { store } from "../../../../src/store";

describe('ItemSearch', () => {
   test('Can search items', () => {
       const component = renderer.create(
           <Provider store={store}>
               <ItemSearch />
           </Provider>
       );

       let tree = component.toJSON();
       expect(tree).toMatchSnapshot();
   });
});
