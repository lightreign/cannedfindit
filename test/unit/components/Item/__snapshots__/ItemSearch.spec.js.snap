// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemSearch Can search items 1`] = `
<form
  className=""
  id="searchForm"
  onSubmit={[Function]}
>
  <div>
    <label
      className="form-label"
      htmlFor="searchItemProductType"
    >
      Search
    </label>
    <input
      className="form-control"
      id="searchItemProductType"
      name="search"
      onChange={[Function]}
    />
  </div>
  <button
    className="btn btn-primary"
    disabled={true}
    type="submit"
  >
    Search
  </button>
  <button
    className="btn btn-secondary"
    disabled={false}
    onClick={[Function]}
    type="submit"
  >
    Clear
  </button>
</form>
`;
